api/internal/init.js

```diff
@@ -1,93 +1,65 @@
   'This could lead to undefined behavior.'
 )
-globalThis.__RUNTIME_INIT_NOW__ = performance.now()
-
-import { IllegalConstructor, InvertedPromise } from '../util.js'
-import { applyPolyfills } from '../polyfills.js'
-import globals from './globals.js'
-import hooks from '../hooks.js'
-import ipc from '../ipc.js'
+console.assert(
+  !globalThis.process?.versions?.node,
+  'socket:internal/init.js was imported in Node.js. ' +
+  'This could lead to undefined behavior.'
+)
 
-import '../console.js'
+globalThis.__RUNTIME_INIT_NOW__ = performance.now()
 
 const GlobalWorker = globalThis.Worker || class Worker extends EventTarget {}
-const hardwareConcurrency = globalThis.navigator?.hardwareConcurrency || 4
 
-const isWorker = !globalThis.window && globalThis.self && globalThis.postMessage
+// only patch a webview or worker context
+if ((globalThis.window || globalThis.self) === globalThis) {
+  if (typeof globalThis.queueMicrotask === 'function') {
+    const originalQueueMicrotask = globalThis.queueMicrotask
+    const promise = Promise.resolve()
 
-class ConcurrentQueue extends EventTarget {
-  concurrency = hardwareConcurrency
-  pending = []
+    globalThis.queueMicrotask = queueMicrotask
+    globalThis.addEventListener('queuemicrotaskerror', (e) => {
+      throw e.error
+    })
 
-  constructor (concurrency) {
-    super()
-    if (typeof concurrency === 'number' && concurrency > 0) {
-      this.concurrency = concurrency
-    }
-  }
+    function queueMicrotask (...args) {
+      if (args.length === 0) {
+        throw new TypeError('Not enough arguments')
+      }
 
-  async wait () {
-    if (this.pending.length < this.concurrency) return
-    await this.peek()
-  }
+      const [callback] = args
 
-  peek () {
-    return this.pending[0]
-  }
+      if (typeof callback !== 'function') {
+        throw new TypeError(
+          'Argument 1 (\'callback\') to globalThis.queueMicrotask ' +
+          'must be a function'
+        )
+      }
 
-  timeout (request, timer) {
-    const onresolve = () => {
-      clearTimeout(timeout)
-      queueMicrotask(() => {
-        const index = this.pending.indexOf(request)
-        if (index > -1) {
-          this.pending.splice(index, 1)
+      return originalQueueMicrotask(task)
+
+      function task () {
+        try {
+          return callback.call(globalThis)
+        } catch (error) {
+          // XXX(@jwerle): `queueMicrotask()` is broken in WebKit WebViews
+          // If an error is thrown, it does not bubble to the `globalThis`
+          // object, but is instead silently discarded. This is misleading
+          // and results in confusion for developers trying to debug something
+          // they may have done wrong. Here we will rethrow the exception out
+          // of microtask execution context with the best function we can
+          // possibly use to report the exception to the `globalThis` object
+          // as an `Unhandled Promise Rejection` error
```

Comment:
👍 

I think we should report the bug on the https://bugs.webkit.org, but I'm not sure what's the minimal repro is

Reply:
oh, found it in the microtask tests